geps/gep-1713.md

```diff
@@ -0,0 +1,220 @@
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+
+The spec currently has language to indicate implementations `MAY` merge Gateways resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## API
+
+We propose adding a new `infrastructure` stanza to the `spec` field of a Gateway. Within `infrastructure` there is an `attachTo` field of type `LocalObjectReference`. Although the use of `LocalObjectReference` allows users to attach to any `kind`, this GEP only defines the behaviour of attaching a Gateway to another Gateway.
+
+
+See [GEP-1867](https://github.com/kubernetes-sigs/gateway-api/pull/1868) for more use cases of `infrastructure`.
+
+#### Go
+
+```go
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+type GatewayInfrastructure struct {
+  // AttachTo allows the Gateway to associate itself with another resource.
+  // A common scenario is to reference another Gateway which marks
+  // this Gateway a child of another.
+  AttachTo LocalObjectReference `json:"attachTo"`
+}
+```
+
+#### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+  labels:
+    gateway.networking.k8s.io/parent-gateway: parent-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: metrics
+    port: 8080
+    protocol: HTTP
+```
+
+### Semantics
+
+
+#### Gateway Attaching
+
+Gateways that _do not_ reference another Gateway (using `spec.infrastructure.attachTo`) are considered "primary" Gateways. A "child" Gateway is one where `spec.infrastructure.attachTo` field is set to a "primary" Gateway. A "sibling" is a Gateway that shares a "parent" with another "child" Gateway.
+
+
+Failure to attach a "child" Gateway to another MUST result in the "child" Gateway setting their `Accepted` condition to `False`. This MUST happen in the following scenarios:
+- Attaching to another "child" Gateway
+- Attaching to oneself (self-referential)
+- Attaching to a non-existing Gateway
+- Attaching to a Gateway with a different `GatewayClassName` 
```

Comment:
I would add "Setting fields only valid on parent Gateways".

At this time, I think that is only spec.addresses. In the future, other parts of `infrastructure`

This could be done in the webhook as well, though (same with self-referential)

Reply:
actually this is mentioned below, just not that it results in Accepted=False

Reply:
> This could be done in the webhook as well, though

I split out the obvious invalid scenarios where a webhook should reject the create/update

> I would add "Setting fields only valid on parent Gateways". ... actually this is mentioned below, just not that it results in Accepted=False

Updated the lower section to mention `Accepted=False`